.github/workflows/tests.yaml

```diff
@@ -47,21 +47,36 @@ jobs:
       # the next job.
       #
       # via https://github.community/t/how-to-use-env-with-container-image/17252/25
-      - name: Save the Docker image name to a variable so we can share and re-use it in other jobs via `${{ needs.deploy-test-synapse-image.outputs.docker_image_name }}`
+      - name: Save the Docker image name to a variable so we can share and re-use it in other jobs via `${{ needs.build-test-synapse-image.outputs.docker_image_name }}`
         id: save_var
-        run: echo "::set-output name=docker_image_name::${{ env.REGISTRY }}/${{ env.GHCR_NAMESPACE }}/${{ env.IMAGE_NAME }}:${{ github.sha }}"
+        run: echo "::set-output name=docker_image_name::${{ env.REGISTRY }}/${{ env.GHCR_NAMESPACE }}/${{ env.IMAGE_NAME }}"
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@98669ae865ea3cffbcbaa878cf57c20bbf1c6c38
+        with:
+          images: ${{ steps.save_var.outputs.docker_image_name }}
+          # Defaults (as indicated by https://github.com/docker/metadata-action/tree/97c170d70b5f8bfc77e6b23e68381f217cb64ded#tags-input).
+          # Plus custom tags:
+          #  - Full length sha
+          tags: |
+            type=schedule
+            type=ref,event=branch
+            type=ref,event=tag
+            type=ref,event=pr
+            type=sha,format=long
```

Comment:
Re-introducing the `docker/metadata-action` so that we get nice tags for PRs, branches, and tags. It just does the right thing.

We additionally added `type=sha,format=long` the default list of tags so that we can still reference it in `services` below. The only change is that it uses the `sha-xxx` prefix for the commit sha tags.